vpr/src/base/xdc_constraints.cpp

```diff
@@ -0,0 +1,244 @@
+        PartitionRegion pr;
+        pr.add_to_part_region(r);
+        part.set_part_region(pr);
+        PartitionId part_id = this->constraints.add_partition(part);
+
+        /* Constrain the AtomBlock */
+
+        this->constraints.add_constrained_atom(port_block, part_id);
+
+        VTR_LOG("XDC: Assigned PACKAGE_PIN %s = [%d, %d, %d, %d].%d\n",
+                pin_name, package_pin_region.x, package_pin_region.y,
+                package_pin_region.w, package_pin_region.h, package_pin_region.subtile);
+
+        return this->_ret_ok();
+    }
+
+    int _set_property_iostandard(Tcl_Obj* tcl_io_standard, Tcl_Obj* tcl_port) {
+        const char* io_standard = Tcl_GetString(tcl_io_standard);
+        if (tcl_io_standard == nullptr)
+            return this->_ret_error("set_property: iostandard should be a string.");
+        
+        auto* port = tcl_obj_getptr<AtomPortId>(tcl_port);
+        if (port == nullptr)
+            return this->_ret_error("set_property: port_name of IOSTANDARD should have a `port` type.");
+        
+        return this->_do_set_property_iostandard(io_standard, *port);
+    }
+
+    int _do_set_property_iostandard(const char* io_standard, AtomPortId port) {
+        AtomBlockId block = this->netlist.port_block(port);
+        this->netlist.set_block_param(block, "IOSTANDARD", io_standard);
+        VTR_LOG("XDC: Assigned IOSTANDARD %s = %s\n", this->netlist.block_name(block).c_str(), io_standard);
+    }
+
+    int _do_get_ports(const char* pin_name) {
+        AtomBlockId pin_block = this->netlist.find_block(pin_name);
+
+        if (pin_block == AtomBlockId::INVALID()) {
+            std::string available_ports;
+            const auto& pins = this->netlist.blocks();
+            int pins_left = pins.size();
+            for (auto& pin_block_ : pins) {
+                auto ports = this->netlist.block_ports(pin_block_);
+                auto port_it = this->netlist.block_ports(pin_block_).begin();
+                if (port_it != ports.end()) {
+                    available_ports += this->netlist.block_name(pin_block_);
+                    if (pins_left != 1)
+                        available_ports += ", ";
+                }
+                pins_left--;
+            }
+
+            return this->_ret_error("set_property: Can't find ports for `" +
+                                    std::string(pin_name) + "`. Available blocks: " +
+                                    available_ports);
+        }
+        
+        auto ports = this->netlist.block_ports(pin_block);
+        auto port_it = this->netlist.block_ports(pin_block).begin();
+        if (port_it == ports.end())
+            return this->_ret_error("set_property: '" + std::string(pin_name) + "' - no ports found.");
+
+        /* TODO: this assumes there's only one port, which just happens to be true in case of fpga interchange.
+         * If we want to handle it better, we should return a list of port ids instead.
+         */
```

Comment:
I think it is ok assume that there is one and only one port. I'd rather add an assert that verifies that the `block_ports` returned are actually of `size == 1`

Reply:
An assert would be good, but a explaining why it's here and why it's valid should remain imo. The `TODO` might be a bit misleading tho, because this isn't going to need any work, unless we end up having more ports per block in the future.

Reply:
Yeah, sounds good, a brief explanation might be good. The fact is that the get_ports, for our use case at the moment, will only refer to a single PAD.

An improvement to this might be to accept some regex as pin name, which will lead to find more than one port and, for instance, assign the same IOSTANDARD. For now, though, we can keep as a single port only.

Reply:
I think this should be done by the user in TCL. We could replace `AtomPortId` returned by `get_ports` with a list of `AtomPortIds` and make it accept a list of strings instead. This way the user could generate that list using a `regexp` or whatever else in TCL.